### PR TITLE
API comparaison refactoring

### PR DESCRIPTION
**Pull request by scrabsha**
This PR adds a huge chunk of refactoring. We're currently able to compare two functions, and a modification if the argument number changes between the previous and next version.

### CODE REVIEW

File: src/public_api.rs [hunk 1]
```diff
@@ -77,12 +77,48 @@ impl PublicApi {
...
+            ApiItem::Mod(_) => todo!(),
+        }
+    }
+
+    pub(crate) fn generate_changes(prev: ApiItem, next: ApiItem) -> Option<Change> {
+        match (prev, next) {
+            (ApiItem::Fn(prev), ApiItem::Fn(next)) => FnMetadata::generate_changes(prev, next),
+            (ApiItem::Mod(_), ApiItem::Mod(_)) => todo!(),
+            _ => unreachable!("Attempt to generate changes for two different-kinded types"),
+        }
```
Comment (scrabsha) [inline src/public_api.rs:106]: We need to simplify this, perhaps with a macro?

File: src/public_api.rs [hunk 2]
```diff
@@ -77,12 +77,48 @@ impl PublicApi {
...
+    }
+
+    pub(crate) fn kind(&self) -> ApiItemKind {
+        match self {
+            ApiItem::Fn(_) => ApiItemKind::Fn,
+            ApiItem::Mod(_) => todo!(),
+        }
+    }
+
+    pub(crate) fn generate_changes(prev: ApiItem, next: ApiItem) -> Option<Change> {
```
Comment (scrabsha) [inline src/public_api.rs:101]: Need a better name for this function

File: src/public_api/functions.rs [hunk 3]
```diff
@@ -2,29 +2,45 @@ use rustc_span::def_id::DefId;
...
+
+    pub(crate) fn path(&self) -> &str {
+        self.path.as_str()
     }
-}
 
-impl DiagnosticGenerator for FnMetadata {
-    fn def_id(&self) -> DefId {
-        self.0
+    pub(crate) fn generate_changes(prev: FnMetadata, next: FnMetadata) -> Option<Change> {
```
Comment (scrabsha) [inline src/public_api/functions.rs:31]: we need a better name

side note: as this function looks very similar to what `ApiItem::generate_changes` does, is it a good idea to put them in a trait, to show that they expose the same set of properties?

File: src/glue.rs [hunk 4]
```diff
@@ -65,68 +73,140 @@ impl MockedCompiler {
...
+        // we don't need the compiler to actually generate any code.
         Compilation::Stop
     }
 }
 
-fn get_diagnosis(tcx: &TyCtxt) -> AnyResult<ApiCompatibilityDiagnostics> {
-    let (prev, curr) =
-        get_previous_and_current_nums(tcx).context("Failed to get dependencies crate id")?;
-    let comparator = ApiComparator::from_crate_nums(prev, curr, tcx);
+pub struct ChangeSet {
```
Comment (scrabsha) [inline src/glue.rs:111]: the change set should not be defined in the glue file

(nothing should be defined in the glue file :sob:)

File: src/comparator.rs [hunk 5]
```diff
@@ -1,65 +1,199 @@
...
 
-        self.item_removals(tcx, &mut collector);
-        self.item_modifications(tcx, &mut collector);
-        self.item_additions(tcx, &mut collector);
+            Diff::Edition(prev, next) => {
+                let p = prev.path();
 
-        let mut diags = collector.finalize();
-        diags.sort();
+                // The pathes for both the previous and next items should be
```
Comment (o0Ignition0o) [inline src/comparator.rs:86]:
```suggestion
                // The paths for both the previous and next items should be
```

**Review by o0Ignition0o — APPROVED** [review]

**Pull request merged**